File: callgrind-utils/testdata/objskip_seed_underflow.c
// Minimal cross-arch reproduction of the arm64 OFF->ON seeding-underflow
// cascade behind the python_fractal_* test failures (full analysis:
// .agents/docs/arm64-python-seeding-underflow-analysis.md).
//
// Call shape, distilled from `python3 -X perf` + ctypes:
//
//   main -> trampoline_call            [asm: FP chain maintained, NO CFI]
//     -> skip_drive                    [obj-skipped .so, "eval loop"]
//       -> skip_begin_hop1 -> skip_begin_hop2      [obj-skipped "ffi dive"]
//         -> clg_begin_marker          [this binary: START fires here]
//       <- returns climb back through the skipped hops
//       -> workload -> leaf_mix        [this binary: the measured region]
//       -> skip_end_hop1 -> skip_end_hop2 -> clg_end_marker   [STOP]
//
// The asm trampoline mimics CPython's -X perf JIT trampolines: it maintains
// the frame-pointer chain but has no .eh_frame FDE. Valgrind's aarch64
// unwinder is CFI-only (m_stacktrace.c), so the OFF->ON seed stops AT the
// trampoline and the seeded context stack is exactly
// [trampoline_call, clg_begin_marker] — one entry deep once
// clg_begin_marker's frame pops. On the next return, bbcc.c's underflow test
// misreads the fn-stack base sentinel as a signal-separation marker,
// handleUnderflow ignores fn->skip, and the skipped hops leak into the graph
// as named, inverted, full-cost nodes. On x86_64 the unwinder's FP fallback
// walks past the trampoline into main/libc, the context stack stays deeper
// than one, and the output is clean — the correct shape on every arch:
//
//   trampoline_call;workload;leaf_mix     (skipped frames folded away)
#include <callgrind.h>
#include <limits.h>
#include <stdlib.h>

extern int skip_drive(int n);
int trampoline_call(int n);

// CPython-trampoline-shaped hop: frame record maintained (so the x86_64
// FP-fallback unwinder can walk through it), but no .cfi_* directives, so no
// FDE is emitted and the CFI-only aarch64 unwinder must stop here.
#if defined(__aarch64__)
__asm__(
    ".text\n"
    ".globl trampoline_call\n"
    ".type trampoline_call, %function\n"
    "trampoline_call:\n"
    "    stp x29, x30, [sp, #-16]!\n"
    "    mov x29, sp\n"
    "    bl skip_drive\n"
    "    ldp x29, x30, [sp], #16\n"
    "    ret\n"
    ".size trampoline_call, .-trampoline_call\n");
#elif defined(__x86_64__)
__asm__(
    ".text\n"
    ".globl trampoline_call\n"
    ".type trampoline_call, @function\n"
    "trampoline_call:\n"
    "    pushq %rbp\n"
    "    movq %rsp, %rbp\n"
    "    call skip_drive@PLT\n"
    "    popq %rbp\n"
    "    ret\n"
    ".size trampoline_call, .-trampoline_call\n");
#else
#error "objskip_seed_underflow: unsupported architecture"
#endif

// Innermost frame of the OFF->ON transition, reached through the skipped
// hops — the clg_start() twin.
__attribute__((noinline)) int clg_begin_marker(void) {
    CALLGRIND_START_INSTRUMENTATION;
    return 1;
}

__attribute__((noinline)) int clg_end_marker(void) {
    CALLGRIND_STOP_INSTRUMENTATION;
    return 1;
}

__attribute__((noinline)) int leaf_mix(int v) {
    return (int)(((unsigned)v * 2654435761u) >> 16);
}

// The measured region, called from the skipped library — must fold as a
// direct child of trampoline_call.
__attribute__((noinline)) int workload(int n) {
    int acc = 1;
    for (int i = 0; i < n; i++)
        acc += leaf_mix(acc + i);
    return acc;
}

int main(int argc, char **argv) {
    // argv[1] = path of the companion .so; register by realpath, exactly as
    // pytest-codspeed / fractal.py do (Callgrind keys obj-skip on the mapped
    // object path).
    char resolved[PATH_MAX];
    if (argc > 1 && realpath(argv[1], resolved))
        CALLGRIND_ADD_OBJ_SKIP(resolved);
    int r = trampoline_call(512);
    return r > 0 ? 0 : 1;
}

File: callgrind-utils/tests/objskip_seed_underflow.rs
//! Minimal cross-arch reproduction of the arm64 OFF->ON seeding-underflow
//! cascade (analysis: `.agents/docs/arm64-python-seeding-underflow-analysis.md`).
//!
//! `testdata/objskip_seed_underflow.c` starts instrumentation two obj-skipped
//! frames below an asm trampoline that maintains the FP chain but carries no
//! CFI — the exact shape of CPython's `-X perf` JIT trampolines in the
//! `python_fractal_*` tests. Correct behavior on every arch: the `skip_*`
//! frames fold away and `workload` parents under `trampoline_call`.
//!
//! On aarch64 the CFI-only unwinder seeds a depth-1 context stack, bbcc.c's
//! underflow heuristic misreads the fn-stack base sentinel as a signal
//! marker, and `handleUnderflow` mints named nodes for `skip=1` functions —
//! these assertions fail there until that is fixed. On x86_64 the FP-fallback
//! unwinder seeds main/libc below the trampoline, the depth never reaches 1,
//! and the assertions pass.
//!
//! Structural assertions only (no insta golden): the contract is
//! platform-independent, and the folded text stays free of libc/arch noise.
use std::env::consts::ARCH;
use std::io::Cursor;
use std::path::{Path, PathBuf};
use std::process::Command;

use callgrind_utils::model::CallGraph;

/// Repo root: this crate lives at `<repo>/callgrind-utils`.
fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("crate has a parent directory")
        .to_path_buf()
}

fn vg_in_place() -> PathBuf {
    let path = repo_root().join("vg-in-place");
    assert!(
        path.is_file(),
        "vg-in-place not found at {} - build Valgrind in place first",
        path.display()
    );
    path
}

/// Compile the skipped companion `.so` and the main binary (which links it
/// directly, with an rpath back to the work dir).
fn compile(work: &Path) -> (PathBuf, PathBuf) {
    let repo = repo_root();
    let testdata = Path::new(env!("CARGO_MANIFEST_DIR")).join("testdata");
    std::fs::create_dir_all(work).expect("create work dir");

    let lib = work.join("libobjskip_seed_underflow.so");
    let status = Command::new("cc")
        // -z now: resolve PLTs eagerly so lazy-binding _dl_runtime_resolve
        // chains don't show up as first-call noise in the folded graph.
        .args(["-g", "-O0", "-shared", "-fPIC", "-Wl,-z,now", "-o"])
        .arg(&lib)
        .arg(testdata.join("objskip_seed_underflow_lib.c"))
        .status()
        .unwrap_or_else(|e| panic!("failed to spawn cc for skip lib: {e}"));
    assert!(status.success(), "cc failed for skip lib ({status})");

    let bin = work.join("objskip_seed_underflow");
    let status = Command::new("cc")
        // -rdynamic: the skipped .so resolves workload/clg_*_marker back in
        // the executable at load time.
        .args(["-g", "-O0", "-rdynamic", "-Wl,-z,now"])
        .arg("-I")
        .arg(repo.join("callgrind"))
        .arg("-I")
        .arg(repo.join("include"))
        .arg("-o")
        .arg(&bin)
        .arg(testdata.join("objskip_seed_underflow.c"))
        .arg(&lib)
        .arg(format!("-Wl,-rpath,{}", work.display()))
        .status()
        .unwrap_or_else(|e| panic!("failed to spawn cc for fixture: {e}"));
    assert!(status.success(), "cc failed for fixture ({status})");
    (bin, lib)
}

/// Profile with the exact runner flag set used by the other fixture tests.
/// The lib path is argv[1]: the fixture realpaths it into
/// `CALLGRIND_ADD_OBJ_SKIP` before instrumentation starts.
fn run_callgrind(bin: &Path, lib: &Path, out_file: &Path) -> String {
    let log_file = out_file.with_extension("valgrind.log");
    let args = [
        "-q",
        "--trace-children=yes",
        "--cache-sim=yes",
        "--I1=32768,8,64",
        "--D1=32768,8,64",
        "--LL=8388608,16,64",
        "--instr-atstart=no",
        "--collect-systime=nsec",
        "--read-inline-info=yes",
        "--tool=callgrind",
        "--compress-strings=no",
        "--combine-dumps=yes",
        "--dump-line=no",
    ];
    let status = Command::new("setarch")
        .arg(ARCH)
        .arg("--addr-no-randomize")
        .arg(vg_in_place())
        .args(args)
        .arg(format!("--callgrind-out-file={}", out_file.display()))
        .arg(format!("--log-file={}", log_file.display()))
        .arg(bin)
        .arg(lib)
        .status()
        .unwrap_or_else(|e| panic!("failed to spawn setarch/vg-in-place: {e}"));
    assert!(status.success(), "vg-in-place exited with {status}");
    std::fs::read_to_string(out_file).unwrap_or_else(|e| panic!("read {}: {e}", out_file.display()))
}

#[test]
fn objskip_seed_underflow_folds_skipped_frames() {
    let work = Path::new(env!("CARGO_TARGET_TMPDIR")).join("objskip_seed_underflow");
    let (bin, lib) = compile(&work);
    let out_file = work.join("objskip_seed_underflow.callgrind.out");
    let raw = run_callgrind(&bin, &lib, &out_file);
    let graph = CallGraph::parse(Cursor::new(raw.as_str()))
        .unwrap_or_else(|e| panic!("parse callgrind output: {e:?}"));
    let folded = graph.to_folded_without_costs();
    let dump = folded.join("\n");

    // 1. Obj-skipped frames must never appear as named nodes. On broken
    //    arm64 the underflow cascade mints skip_begin_hop1/_ctypes-style
    //    inverted nodes for skip=1 functions.
    let leaked: Vec<&String> = folded.iter().filter(|l| l.contains("skip_")).collect();
    assert!(
        leaked.is_empty(),
        "obj-skipped frames leaked into the folded graph \
         (OFF->ON seeding-underflow cascade):\n{}\n\nfull folded output:\n{dump}",
        leaked
            .iter()
            .map(|s| s.as_str())
            .collect::<Vec<_>>()
            .join("\n")
    );

    // 2. The measured region must parent under the trampoline (nearest
    //    non-skipped frame), like py::run_measured;py::complex_fractal_benchmark.
    assert!(
        folded
            .iter()
            .any(|l| l.starts_with("trampoline_call;workload")),
        "workload is not parented under trampoline_call:\n{dump}"
    );

    // 3. Roots: only the seeded innermost frame (clg_begin_marker, tiny
    //    post-request residue) and the trampoline may be roots.
    for line in &folded {
        let root = line
            .split(';')
            .next()
            .unwrap()
            .trim_end_matches(" <cost>");
        assert!(
            root == "trampoline_call" || root == "clg_begin_marker",
            "unexpected root {root:?} in folded graph:\n{dump}"
        );
    }
}
